curvefs/src/mds/schedule/leaderScheduler.cpp

```diff
@@ -0,0 +1,387 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * @Project: curve
+ * @Date: 2022-04-07 14:15:49
+ * @Author: chenwei
+ */
+
+#include "curvefs/src/mds/schedule/operatorFactory.h"
+#include "curvefs/src/mds/schedule/scheduler.h"
+
+namespace curvefs {
+namespace mds {
+namespace schedule {
+/**
+ * use curl -L mdsIp:port/flags/enableRapidLeaderScheduler?setvalue=true
+ * for dynamic parameter configuration
+ *
+ * If this value is set true, ignore metaserverCoolingTimeSec_
+ * when choose metaserver as a leader when leader schedule.
+ *
+ * if this value is set true, when no leader schedule operation at this round,
+ * set the value to false automaticlly.
+ */
+static bool pass_bool(const char *, bool) { return true; }
+DEFINE_bool(enableRapidLeaderScheduler, false, "rapid leader scheduler once");
+DEFINE_validator(enableRapidLeaderScheduler, &pass_bool);
+
+int LeaderScheduler::Schedule() {
+    LOG(INFO) << "LeaderScheduler begin";
+
+    int oneRoundGenOp = 0;
+    for (auto poolId : topo_->Getpools()) {
+        oneRoundGenOp += LeaderSchedulerForPool(poolId);
+    }
+
+    LOG(INFO) << "LeaderScheduler generate " << oneRoundGenOp
+              << " operators at this round";
+
+    if (FLAGS_enableRapidLeaderScheduler && oneRoundGenOp == 0) {
+        LOG(INFO) << "LeaderScheduler enable rapid leader scheduler and "
+                  << " generate no opration in this round, "
+                  << "set enableRapidLeaderScheduler to false";
+        FLAGS_enableRapidLeaderScheduler = false;
```

Comment:
Why set enableRapidLeaderScheduler to false when no leader schedule operation at this round?

Reply:
默认不进行rapid schedule，如果通过命令设置了rapid schedule，就会按照rapid 的逻辑去均衡。当没有新的operation，就可以认为这次rapid schedule结束了。
The rapid schedule is not performed by default. If the rapid schedule is set through the command, it will be balanced according to the logic of the rapid. When there is no new operation, it can be considered that the rapid schedule is over.